### PR TITLE
Adding handling for expired intermediate certs when verifying client

### LINKED ISSUE #787: Expired intermediate certificates prevents maintaining connections
### Describe the bug

@CMCDragonkai was having a problem where his node could not connect to any other nodes. Even weirder he could ping the seed nodes but these connections would not linger in the `NodeConnectionManager`. While this is happening we're getting the following error.

```
INFO:polykey.PolykeyAgent.NodeConnectionManager.NodeConnectionForward [13.239.117.143:1314].QUICClient.QUICConnection 6a6cded032c25659a375b3959654b94eed9ef41c:ErrorQUICConnectionPeerTLS: QUIC Connection local TLS error - Peer closed with transport code 301
INFO:polykey.PolykeyAgent.NodeConnectionManager.NodeConnectionForward [13.239.117.143:1314].QUICClient:ErrorQUICConnectionPeerTLS: QUIC Connection local TLS error - Peer closed with transport code 301
```

Digging into the codes `301` is the error for an expired certificate. Looking deeper into the verification logic it seems to be trying to verify all of the certificates within the chain.

### To Reproduce

1. create a node.
2. wait for the node's certificate to expire
3. Rotate the certificate
4. attempt any connections to other nodes
5. Any connections will establish but be force closed immediately when the client verification fails.

### Expected behaviour

We should be able to make and maintain connections even if the intermediate certificates fail to validate.

### Screenshots

Certificates:

```
»» ~/Projects/Polykey-CLI/tmp
 ♖ step certificate inspect ./cert1                          (staging) pts/2 15:23:15
Certificate:
    Data:
        Version: 3 (0x2)
        Serial Number: 8495915809015553842318195064004662698 (0x66440fd5f5c7000b1595f8a5ccfd5aa)
    Signature Algorithm: Ed25519
        Issuer: CN=vve3f83v1qog578bgav365bmmgtt71prpdogs2f2dcqcf59i60br0
        Validity
            Not Before: May 15 01:28:53 2024 UTC
            Not After : May 15 01:28:53 2025 UTC
        Subject: CN=vve3f83v1qog578bgav365bmmgtt71prpdogs2f2dcqcf59i60br0
        Subject Public Key Info:
            Public Key Algorithm: Ed25519
                Public-Key: (32 bit)
                    fb:86:f4:0f:e1:d6:20:53:a1:70:57:c6:62:ae:d6:
                    87:7a:70:e7:79:6e:21:c1:3c:4d:66:98:f2:a6:46:
                    02:f6
        X509v3 extensions:
            X509v3 Basic Constraints:
                CA:TRUE
            X509v3 Key Usage:
                Digital Signature, Content Commitment, Key Encipherment, Data Encipherment, Key Agreement, Certificate Sign, CRL Sign
            X509v3 Extended Key Usage:
                Server Authentication, Client Authentication, Code Signing, E-mail Protection, Time Stamping, OCSP Signing
            X509v3 Subject Alternative Name:
                DNS:vve3f83v1qog578bgav365bmmgtt71prpdogs2f2dcqcf59i60br0
                IP Address:127.0.0.1, IP Address:::1
                URI:pk://vve3f83v1qog578bgav365bmmgtt71prpdogs2f2dcqcf59i60br0
            X509v3 Subject Key Identifier:
                15:79:EB:AD:79:93:CC:A6:F1:2F:59:90:FF:E8:83:46:43:CE:20:49
            1.3.6.1.4.1.57167.2.2.1:
                ..1.4.0
            1.3.6.1.4.1.57167.2.2.2:
                .@........X....2F..bɚT....^?....2.Z.1....\..Z.φ́.|1..gD..Ahc...
    Signature Algorithm: Ed25519
         c4:fc:cd:1b:45:28:a2:0b:b0:b2:26:65:cf:d3:b1:0b:11:29:
         95:ab:d9:a5:14:17:47:31:45:52:b4:43:12:ed:59:f2:4b:d6:
         59:02:b9:78:1d:a7:2c:47:92:8b:89:d0:35:f8:fd:df:77:1e:
         0e:0c:92:43:f7:60:06:11:82:04
»» ~/Projects/Polykey-CLI/tmp
 ♖ step certificate inspect ./cert2                          (staging) pts/2 15:26:36
Certificate:
    Data:
        Version: 3 (0x2)
        Serial Number: 8296194186598527926257437087922070292 (0x63dc9f418497000a1ae98c2ceb32314)
    Signature Algorithm: Ed25519
        Issuer: CN=vve3f83v1qog578bgav365bmmgtt71prpdogs2f2dcqcf59i60br0
        Validity
            Not Before: Feb 3 05:44:33 2023 UTC
            Not After : Feb 3 05:44:33 2024 UTC
        Subject: CN=vve3f83v1qog578bgav365bmmgtt71prpdogs2f2dcqcf59i60br0
        Subject Public Key Info:
            Public Key Algorithm: Ed25519
                Public-Key: (32 bit)
                    fb:86:f4:0f:e1:d6:20:53:a1:70:57:c6:62:ae:d6:
                    87:7a:70:e7:79:6e:21:c1:3c:4d:66:98:f2:a6:46:
                    02:f6
        X509v3 extensions:
            X509v3 Basic Constraints:
                CA:TRUE
            X509v3 Key Usage:
                Digital Signature, Content Commitment, Key Encipherment, Data Encipherment, Key Agreement, Certificate Sign, CRL Sign
            X509v3 Extended Key Usage:
                Server Authentication, Client Authentication, Code Signing, E-mail Protection, Time Stamping, OCSP Signing
            X509v3 Subject Alternative Name:
                DNS:vve3f83v1qog578bgav365bmmgtt71prpdogs2f2dcqcf59i60br0
                IP Address:127.0.0.1, IP Address:::1
                URI:pk://vve3f83v1qog578bgav365bmmgtt71prpdogs2f2dcqcf59i60br0
            X509v3 Subject Key Identifier:
                15:79:EB:AD:79:93:CC:A6:F1:2F:59:90:FF:E8:83:46:43:CE:20:49
            1.3.6.1.4.1.57167.2.2.1:
                ..1.0.1-alpha.0
            1.3.6.1.4.1.57167.2.2.2:
                .@.ީ...q.^........c...j7.S..|A.C.\m; i......wf..$.f.J..ߤ.h...t..
    Signature Algorithm: Ed25519
         6b:dc:24:48:29:87:e8:7e:a1:72:ed:18:ce:bb:93:68:a8:77:
         35:72:ce:28:ed:43:3f:d6:3e:7d:4a:f0:3f:8e:a0:83:97:85:
         fd:3e:fa:e3:41:60:eb:4d:24:5c:96:01:06:30:c5:2c:2f:f7:
         2e:b9:9b:0a:e6:22:65:b9:5e:0a
```

### PR DESCRIPTION
**Pull request by tegefaulkes**
### Description

This PR fixes a bug with expired intermediate certificates preventing connections when connecting with a client. 

### Issues Fixed

Fixes #787

### Tasks

- [X] 1. Fix bug in `verifyClientCertificateChain` that prevented connections when we had an expired certificate when the chain should still be valid.
- [X] 2. Check `verifyServerCertificateChain` for any similar issues.
- [X] 3. Expand tests to check for this edge case with expired intermediate certificates

### Final checklist

* [x] Domain specific tests
* [x] Full tests
* [x] Updated inline-comment documentation
* [x] Lint fixed
* [x] Squash and rebased
* [x] Sanity check the final build

### CODE REVIEW

**Comment by tegefaulkes** [comment]
I should add 1 more test for a chain that is `[valid-leaf, expired, valid-target, root]`.

Also @aryanjassal have a quick review of this just to get an idea of the change.

**Comment by CMCDragonkai** [comment]
We need a new release for this for my node to continue working.

File: src/nodes/utils.ts [hunk 1]
```diff
@@ -647,10 +647,11 @@ async function verifyClientCertificateChain(
     certChain.push(cert);
   }
   const now = new Date();
+  let leafCert = true;
   for (let certIndex = 0; certIndex < certChain.length; certIndex++) {
     const cert = certChain[certIndex];
     const certNext = certChain[certIndex + 1];
-    if (now < cert.notBefore || now > cert.notAfter) {
+    if (leafCert && (now < cert.notBefore || now > cert.notAfter)) {
```
Comment (CMCDragonkai) [inline src/nodes/utils.ts:654]: This is a bit confusing when it's set to true and then checked. The order of switching booleans may be clearer if it was the other way around.

Comment (tegefaulkes) [reply src/nodes/utils.ts:654]: As in, started as false and had a different name?

**Comment by CMCDragonkai** [comment]
Starting something as true and then checking if it is true is just a bit weird even if you're later switching it to false. In fact it sounds like you should be doing an initial step on the head and then loop over the tail. These should be 2 different procedures.

23 Aug 2024 10:45:36 Brian Botha ***@***.***>:
